feat: SEPA Payment Order

barredterra opened the pull request:
Create SEPA Credit Transfer from ERPNext. Download as XML file or directly upload to bank.

Changes:

- New DocType **SEPA Payment Order**
    - Validates all IBANs
    - Automatically sets BIC and Bank Name for German IBANs
    - Validates payment currency against bank account currency (_Company Account_ linked in **Bank Account**)
    - Sets _Transmission Type_ and _Transmission Datetime_ on each export of payment data. Changes are tracked.
    - Submit -> button "Download as XML" provides a pain.001 file that can be manually uploaded to the bank
        Sets _Transmission Type_ "DOWNLOADED".
    - Submit -> button "Send to Bank" uploads data to bank via EBICS.
        Sets _Transmission Type_ "SENT_VIA_EBICS".
        Note: This is unsigned, it still needs to be signed by an authorized user before it is executed.
- Create a mapped **SEPA Payment Order** from one or many **Purchase Invoices**
    - Form view -> Create -> SEPA Payment Order
    - List view -> Select records -> Actions -> SEPA Payment Order
    - Maps one or many **Payment Schedule** rows to one or many **SEPA Payment** rows in a **SEPA Payment Order**
    - Uses the _Supplier Bank Account_ selected in **Purchase Invoice**, or the (default) **Bank Account** linked to the **Supplier**.
- Adjusted the banking workspace

Todo:

- [x] Create mapped **Bank Account** from **Supplier** https://github.com/frappe/erpnext/pull/49000
- [x] Allow to manually set transmission date
- [x] Find a good way to link SEPA Payment Orders to reference transactions. -> Dynamic Link from payment to transaction
- [x] How do I know that a payment has been ordered for a certain invoice: Status field in Payment Schedule. Visible only in Purchase Invoice. Link schedule name in payment.
- [x] Quick create **Bank Account** from **Purchase Invoice**: Button, ask for IBAN, the rest can be determined automatically.
- [x] Hide Passphrase field when it's stored in EBICS User

Ref: DRC-158

barredterra committed:
feat: SEPA Payment Order

Create SEPA Credit Transfer from ERPNext. Download as XML file or directly upload to bank.

barredterra committed:
feat: create SEPA Payment Order from Purchase Invoice

barredterra committed:
fix(SEPA Payment): adjust column width for IBAN

barredterra committed:
fix: adjust banking workspace

- added SEPA Payment Order
- Cards "Setup" and "Transactions"

barredterra committed:
feat(SEPA Payment Order): validate IBANs

barredterra committed:
feat(SEPA Payment Order): set bank name and swift number from german IBAN

PatrickDEissler commented on banking/custom/purchase_invoice.js:
```diff
@@ -0,0 +1,17 @@
+frappe.ui.form.on("Purchase Invoice", {
+	refresh(frm) {
+		if (flt(frm.doc.outstanding_amount, 2) > 0.0) {
```
```suggestion
		if (frm.doc.docstatus == 1 && flt(frm.doc.outstanding_amount, 2) > 0.0) {
```

PatrickDEissler commented on banking/custom/purchase_invoice.py:
```diff
@@ -0,0 +1,69 @@
+
+
+@frappe.whitelist()
+def make_sepa_payment_order(source_name: str, target_doc=None):
```
Is there no extra permission-check needed for this whitelisted method? 
Is the permission-check already included `get_mapped_doc`?

barredterra replied:
Yes, `get_mapped_doc` takes care of permissions

PatrickDEissler commented on banking/custom/purchase_invoice.py:
```diff
@@ -0,0 +1,69 @@
+	)
+
+
+@frappe.whitelist()
+def make_bulk_sepa_payment_order(source_names: str):
```
No permission-check needed in whitelisted method?

barredterra replied:
`get_mapped_doc` takes care of permissions

barredterra committed:
feat: store latest transmission type and date

- for download as well
- changes will be logged
- transmission requires submit perms
- Download via POST request

barredterra committed:
chore: export python type annotations

barredterra committed:
fix: store transmission datetime instead of only date

barredterra committed:
fix(purchase invoice): use same condition in list and form

barredterra committed:
feat: link Payment Order to Purchase Invoice

- Add fields Reference DocType and Reference Name
- Insert link into Purchase Invoice dashboard
- Set make_methods in Purchase Invoice, so Payment Order is created correctly via dashboard link

barredterra committed:
feat(SEPA Payment Order): validate payment currencies against bank account currency

barredterra committed:
fix(SEPA Payment Order): Transmission Type is empty at the beginning

barredterra committed:
feat(Purchase Invoice): choose Supplier Bank Account

barredterra committed:
fix: set reference doc in SEPA Payment created from Purchase Invoice

barredterra committed:
refactor: move custom fields to separate file

barredterra committed:
fix(SEPA Payment): store Reference Row Name

barredterra committed:
feat(SEPA Payment Order): trigger hook on reference docs when status changes

barredterra committed:
feat(Purchase Invoice): store Payment Order status in Payment Schedule row

barredterra committed:
feat(Purchase Invoice): create Supplier Bank Account from IBAN

barredterra committed:
fix(SEPA Payment Order): hide Passprase field if stored in EBICS User

barredterra committed:
fix: use order type FUL for upload without execution

barredterra committed:
fix: remove spaces in IBAN

barredterra committed:
chore(SEPA Payment Order): temporarily disable "Send to Bank"

It's not fully tested yet and will be activated later

barredterra committed:
Merge remote-tracking branch 'upstream/version-15-hotfix' into drc-158

barredterra commented:
@Mergifyio backport version-14-hotfix
